src/core/ext/filters/client_channel/subchannel_index.c

```diff
@@ -119,13 +121,23 @@ static const gpr_avl_vtable subchannel_avl_vtable = {
 void grpc_subchannel_index_init(void) {
   g_subchannel_index = gpr_avl_create(&subchannel_avl_vtable);
   gpr_mu_init(&g_mu);
+  gpr_ref_init(&g_refcount, 1);
 }
 
 void grpc_subchannel_index_shutdown(void) {
-  grpc_exec_ctx exec_ctx = GRPC_EXEC_CTX_INIT;
-  gpr_mu_destroy(&g_mu);
-  gpr_avl_unref(g_subchannel_index, &exec_ctx);
-  grpc_exec_ctx_finish(&exec_ctx);
+  grpc_subchannel_index_unref();
```

Comment:
The goal of this PR is to not consider the client_channel plugin to be shut down until after all resolver and LB policy callbacks are invoked.  However, I don't think this actually does that; it just defers destroying the subchannel index's global mutex until the LB policy is destroyed -- but if there are still pending callbacks to be invoked, that will never actually happen unless we do some polling to kick them off.

Does the iomgr shutdown do polling to make sure everything is shut down?  If not, the net effect of this will be to replace a crash with a memory leak.  This may be a slight improvement, but it's certainly not ideal. :)

If the iomgr shutdown doesn't do any polling, then I think we may need to do something like this:

```
while (!gpr_ref_is_unique(&g_refcount)) {
  // ...do some polling to give the pending LB callbacks a chance to be executed...
}
grpc_subchannel_index_unref();
```

I'm not sure what the best way is to get the polling to happen here.  David or Craig or Yang might have some ideas.

Reply:
I don't fully understand it.

1. If the callbacks hold ref's to the LB policy and we destroyed the LB policy, doesn't that mean the callbacks have all been invoked? Can the LB policy be destroyed without the callbacks invocation after some timeout?
2. I made every callback hold a ref to the subchannel_index because I assume every callback will be eventually invoked, so the last one will destroy the subchannel_index resource if any one is left over after the regular shutdown. Actually not every callback will be invoked if the shutdown is finished quickly?

Reply:
C-core is basically a giant event loop.  Callbacks are generally only invoked as a result of polling on file descriptors.  If callbacks have not yet been invoked by the time we shut down C-core, then it is likely that they will never be invoked unless some polling is done.

This change fixes the crash by ensuring that we don't destroy the global mutex until all callbacks have been invoked.  However, if the callbacks are never going to be run, then we will still have a memory leak, because the refcount will never reach zero, so we will never clean up the subchannel index.

(There's also a pre-existing leak caused by the fact that callbacks that are not being run are holding references to the LB policies, which hold references to the channels, so there are a lot of other objects that are not being destroyed.  But the solution is the same for both the pre-existing leaks and the new leaks: We need to make sure that the callbacks actually get invoked.)

Note that the iomgr library is where the polling engine lives, so the latest possible point at which polling can happen is the iomgr shutdown.  It's possible that the iomgr shutdown code tries to do some last-minute polling on its own (which is where the memory leaks are identified for `GRPC_ABORT_ON_LEAKS=true`), in which case there is probably no problem here.  But you should check that, because if it's not doing any polling, then we'll probably need to do it here.

Reply:
I don't find any polling in `grpc_iomgr_shutdown()`. It does call a function called `grpc_iomgr_platform_flush()` but its body is empty. 

Can we assume there won't be any new callbacks (especially those using `subchannel_index`) created after we destroy `subchannel_index`?

I am not sure how to do polling in `grpc_subchannel_index_shutdown()`. I will ask Sree tomorrow.

Reply:
@sreecha @dgquintas This is where we should flush the callbacks. Any idea how to do it properly? Thanks!